### ISSUE TITLE
False positive for react/display-name on function value assignment

### ISSUE THREAD

**Issue by apbarrero** [jsx-eslint/eslint-plugin-react#3329]
Related to https://github.com/jsx-eslint/eslint-plugin-react/issues/3289 and reproducible after its fix in 7.30.1.

This code makes the `display-name` rule to yield the error `'Component definition is missing display name'`. So it seems that it is still identifying the assignment of a function value to a variable as a component definition.

```
let demo = null;
demo = (a) => {
  if (a == null) return null;
  return f(a);
}`,
```

As opposed to https://github.com/jsx-eslint/eslint-plugin-react/issues/3289, it is needed both that the `demo` variable is declared and assigned in two different statements and that the value returned by the function is not a function itself (i.e. not a higher order function).

**Comment by ljharb** [comment]
Why is that needed?

(Either way, this function is a lowercase-named function so it shouldn't be detected as a component)

**Comment by apbarrero** [comment]
@ljharb well, as you said, it shouldn't be detected as a component and it is valid code, so the rule shouldn't claim it as an error.

Why are you asking?

**Comment by ljharb** [comment]
I’m curious.
